pypeit/core/flat.py

```diff
@@ -1,486 +1,416 @@
-        plt.hlines(sigrej_illum * var_value, ximg_twod.min(), ximg_twod.max(), color='lawngreen', linestyle='--',
-                   label='rejection thresholds', zorder=10,linewidth=2.0)
-        plt.hlines(-sigrej_illum * var_value, ximg_twod.min(), ximg_twod.max(), color='lawngreen', linestyle='--',
-                   zorder=10,linewidth=2.0)
-        ax.set_ylim((-0.05, 0.05))
-        ax.set_xlim(-0.02, 1.02)
-        plt.legend()
-        plt.xlabel('Normalized Slit Position')
-        plt.ylabel('Residuals from pixelflat 2-d fit')
-        plt.title('Spatial Residuals for slit={:d}'.format(slit))
+        ax.scatter(spat_coo_data[spat_gpm_data_raw], spat_flat_data_raw[spat_gpm_data_raw],
+                   marker='.', lw=0, s=10, color='k', zorder=1, label='used data')
+        ax.scatter(spat_coo_data[np.invert(spat_gpm_data_raw)],
+                   spat_flat_data_raw[np.invert(spat_gpm_data_raw)],
+                   marker='.', lw=0, s=10, color='C3', zorder=2, label='rejected data')
+        ax.plot(spat_coo_data[spat_gpm_data_raw], spat_flat_data, color='C2', zorder=3,
+                label='filtered profile')
+        ax.legend()
+        ax.set_title('Optimized slit illumination profile')
+        ax.set_xlabel('Spatial coordinate')
+        ax.set_ylabel('Spectrally collapsed, normalized flux')
         plt.show()
 
-    # Evaluate and save
-    twod_model = np.ones_like(flat)
-    twod_this = np.zeros_like(twodfit)
-    twod_this[isrt_spec] = twodfit
-    twod_model[thismask_out] = twod_this
-
-    # Compute all the final output images output
-    pixelflat = np.ones_like(flat)
-    flat_model = np.ones_like(flat)
-    flat_model[thismask_out] = twod_model[thismask_out]*np.fmax(illumflat[thismask_out],0.05)*np.fmax(spec_model[thismask_out],1.0)
-    pixelflat[thismask_out] = flat[thismask_out]/flat_model[thismask_out]
-
-    # ToDo Add some code here to treat the edges and places where fits go bad?
-    # Set the pixelflat to 1.0 wherever the flat was nonlinear
-    pixelflat[flat >= nonlinear_counts] = 1.0
-    # Do not apply pixelflat field corrections that are greater than 100% to avoid creating edge effects, etc.
-    # TODO Should we do the same for the illumflat??
-    #pixelflat = np.fmax(np.fmin(pixelflat, 2.0), 0.5)
-    pixelflat = np.clip(pixelflat, 0.5, 2.0)
+    # Include the rejected data in the full image good-pixel mask
+    _spat_gpm[_spat_gpm] = spat_gpm_data_raw[np.argsort(spat_srt)]
+    # Recreate the illumination profile data
+    _spat_gpm, spat_srt, spat_coo_data, spat_flat_data_raw \
+            = sorted_flat_data(norm_spec, spat_coo, gpm=_spat_gpm)
+    return _spat_gpm, spat_srt, spat_coo_data, spat_flat_data_raw, \
+                illum_filter(spat_flat_data_raw, med_width)
+
+# TODO: See pypeit/deprecated/flat.py for the previous version. We need
+# to continue to vet this algorithm to make sure there are no
+# unforeseen corner cases that cause errors.
```

Comment:
How much faster is this new algorithm compared to the last one. The brute force approach of marching inward from the left and right is in principle more robust.

Reply:
We talked through this.  It's about an order of magnitude faster, but we're talking about the difference between a few tenths of a second and a few hundredths, so it's not a huge deal.  We decided to continue to use the current algorithm, but just to keep an eye out for any unforeseen failure modes.